apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/source/RdaVersion.java

```
@@ -0,0 +1,172 @@
+
+    /** Empty private constructor to prevent direct instantiation. */
+    private RdaVersionBuilder() {
+      // Nothing to do
+    }
+
+    /**
+     * Sets the version string to parse when building the {@link RdaVersion} object.
+     *
+     * @param versionString The version string to set.
+     * @return A reference to this builder instance.
+     */
+    public RdaVersionBuilder versionString(String versionString) {
+      this.versionString = versionString;
+      return this;
+    }
+
+    /**
+     * Builds an {@link RdaVersion} object instance using the previously given parameters.
+     *
+     * @return An {@link RdaVersion} object built with the previously provided parameters.
+     */
+    public RdaVersion build() {
+      if (versionString != null) {
+        Matcher matcher = VERSION_PATTERN.matcher(versionString);
+
+        CompatibilityLevel level;
+        int major;
+        int minor;
+        int patch;
+
+        if (matcher.matches()) {
+          String compatibility = matcher.group(TYPE_GROUP);
+
+          if (compatibility != null) {
+            level = compatibility.equals("~") ? CompatibilityLevel.MINOR : CompatibilityLevel.MAJOR;
+          } else {
+            level = CompatibilityLevel.PATCH;
+          }
+
+          try {
+            major = Integer.parseInt(matcher.group(MAJOR_GROUP));
+            minor = Integer.parseInt(matcher.group(MINOR_GROUP));
+            patch = Integer.parseInt(matcher.group(PATCH_GROUP));
+          } catch (NumberFormatException e) {
+            // This shouldn't happen
+            throw new IllegalStateException("Failed to parse RDA Version");
+          }
+        } else {
+          throw new IllegalArgumentException(
+              String.format("Invalid RdaVersion format '%s'", versionString));
+        }
+
+        return new RdaVersion(major, minor, patch, level);
+      } else {
+        throw new NullPointerException("versionString can not be null");
+      }
```

🍰 Could reduce nesting by checking this first and just throwing the exception if versionString=null instead of having nesting all the other logic and having this be the else. (Same with the other validations)

---

It's a styling choice, it can be argued that "fast fail" creates isolated path branches that can lead to potential bugs introduced, where "happy-path first" increases nesting depth.  I've always been more partial to the latter.

---

Fair enough; I've always seen the fast fail method pushed for but it is ultimately a stylistic choice.